File: backend/src/main/java/com/travelcompanion/controller/AdminController.java
package com.travelcompanion.controller;

import com.travelcompanion.dto.DestinationDTO;
import com.travelcompanion.dto.TripResponse;
import com.travelcompanion.entity.User;
import com.travelcompanion.repository.UserRepository;
import com.travelcompanion.service.DestinationService;
import com.travelcompanion.service.TripService;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/api/admin")
@RequiredArgsConstructor
public class AdminController {

    private final UserRepository userRepository;
    private final DestinationService destinationService;
    private final TripService tripService;

    @GetMapping("/stats")
    public ResponseEntity<Map<String, Object>> getStats() {
        Map<String, Object> stats = new HashMap<>();
        stats.put("totalUsers", userRepository.count());
        stats.put("totalDestinations", destinationService.getAllDestinations().size());
        stats.put("totalTrips", tripService.getAllTrips().size());
        return ResponseEntity.ok(stats);
    }

    @GetMapping("/users")
    public ResponseEntity<List<Map<String, Object>>> getAllUsers() {
        List<Map<String, Object>> users = userRepository.findAll().stream()
                .map(user -> {
                    Map<String, Object> map = new HashMap<>();
                    map.put("id", user.getId());
                    map.put("username", user.getUsername());
                    map.put("email", user.getEmail());
                    map.put("role", user.getRole().name());
                    map.put("createdAt", user.getCreatedAt());
                    return map;
                })
                .toList();
        return ResponseEntity.ok(users);
    }

    @DeleteMapping("/users/{id}")
    public ResponseEntity<Void> deleteUser(@PathVariable Long id) {
        userRepository.deleteById(id);
        return ResponseEntity.noContent().build();
    }

    @GetMapping("/trips")
    public ResponseEntity<List<TripResponse>> getAllTrips() {
        return ResponseEntity.ok(tripService.getAllTrips());
    }

    @PostMapping("/destinations")
    public ResponseEntity<DestinationDTO> addDestination(@RequestBody DestinationDTO dto) {
        return ResponseEntity.ok(destinationService.createDestination(dto));
    }

    @PutMapping("/destinations/{id}")
    public ResponseEntity<DestinationDTO> updateDestination(@PathVariable Long id, @RequestBody DestinationDTO dto) {
        return ResponseEntity.ok(destinationService.updateDestination(id, dto));
    }

    @DeleteMapping("/destinations/{id}")
    public ResponseEntity<Void> deleteDestination(@PathVariable Long id) {
        destinationService.deleteDestination(id);
        return ResponseEntity.noContent().build();
    }
}

File: backend/src/controllers/admin.controller.js
const adminService = require('../services/admin.service');
const destinationService = require('../services/destination.service');

exports.getStats = async (req, res, next) => {
  try {
    const stats = await adminService.getStats();
    res.json(stats);
  } catch (error) {
    next(error);
  }
};

exports.getAllUsers = async (req, res, next) => {
  try {
    const users = await adminService.getAllUsers();
    res.json(users);
  } catch (error) {
    next(error);
  }
};

exports.deleteUser = async (req, res, next) => {
  try {
    await adminService.deleteUser(req.params.id);
    res.status(204).end();
  } catch (error) {
    if (error.message === 'User not found') {
      return res.status(404).json({ message: error.message });
    }
    if (error.message === 'Cannot delete admin user') {
      return res.status(403).json({ message: error.message });
    }
    next(error);
  }
};

exports.getAllTrips = async (req, res, next) => {
  try {
    const trips = await adminService.getAllTrips();
    res.json(trips);
  } catch (error) {
    next(error);
  }
};

exports.addDestination = async (req, res, next) => {
  try {
    const destination = await destinationService.create(req.body);
    res.status(201).json(destination);
  } catch (error) {
    next(error);
  }
};

exports.updateDestination = async (req, res, next) => {
  try {
    const destination = await destinationService.update(req.params.id, req.body);
    res.json(destination);
  } catch (error) {
    if (error.message === 'Destination not found') {
      return res.status(404).json({ message: error.message });
    }
    next(error);
  }
};

exports.deleteDestination = async (req, res, next) => {
  try {
    await destinationService.delete(req.params.id);
    res.status(204).end();
  } catch (error) {
    if (error.message === 'Destination not found') {
      return res.status(404).json({ message: error.message });
    }
    next(error);
  }
};
